students/meet_shah/lesson05/mailroom3.py

```diff
@@ -0,0 +1,121 @@
+        }
+
+def thank_you():
+    while True:
+        new_response = input("Enter a command. Your options are [list, <donor name>, q(quit),]>>")
+        if new_response == "list":
+            print("\nHere's a list of our donors:\n")
+
+            ### Added list comprehension ###
+            [print(donor) for donor, _ in donors.items()]
+        elif new_response in ['q','quit']:
+            print("Returning to the main menu\n")
+            break
+        else:
+            donor = new_response
+            donation = input("Enter the amount you would like to donate: ")
+            donation = int(donation)
+            donors.setdefault(new_response, []).append(donation)
+            print("Dear {}, We gladly accept your generous donation of ${}".format(donor,donation))
+            print("Please visit us anytime to add more donations")
+            break
+
+
+
+
+def print_report(t):
+    return '{:20}{:>20}{:>12}{:>14}'.format(*t)
+
+
+def report():
+    print("Donor Name                | Total Given | Num Gifts | Average Gift")
+    print("-"*66)
+    reverse_donor = {}
+    for donor, donations in donors.items():
+        total_donations = sum(donations)
+        reverse_donor[total_donations] = donor
+
+
+
+    total_donations = sorted(reverse_donor.keys(), reverse=True)
+    for total_donation in total_donations:
+        donor = reverse_donor[total_donation]
+        num_donations = donors[donor]
+        num_donations = len(donations)
+        average_donations = total_donation * 1.0 / num_donations
+        average_donations = "${:,.2f}".format(average_donations)
+        total_donation = "${:,.2f}".format(total_donation)
+        print(print_report((donor, total_donation, num_donations, average_donations)))
+
+def send_letters():
+    for donor, donations in donors.items():
+        filename = donor.replace(" ", "_") + ".txt"
+        with open(filename, "w") as infile:
+            infile.write("Dear {},\n\n".format(donor))
+            infile.write("\tThank you for your generous donation of {}\n\n".format(donations[-1]))
+            infile.write("\tIt will be put to very good use.\n\n")
+            infile.write("\t\t\t\t\t   Sincerely, \n\t\t\t\t\t   The Team\n")
+
+
+
+
+def quit():
+    print("You have chosen to quit! GoodBye!")
+    sys.exit()
+
+def mainloop():
+    print("Welcome to Mailroom")
+    response = ""
+    while True:
+
+        menu_dict = {1: thank_you, 2: report,
+                     3: send_letters, 4: quit}
+
+        print("What do you want to do? Choose from a menu of 3 actions:")
+        print("1. Send a Thank You")
+        print("2. Create a Report")
+        print("3. Send letters to everyone")
+        print("4. Quit")
+        response = input(">>")
+        print("Your response was:", response)
+
+        ### added try/except clause
+        try:
+            myresponse = int(response)
+        except ValueError:
+            continue
+
+        try:
```

Comment:
These two are a little repetitive, and you only really need the 2nd one. I guess you could have done a different try/except in the thank_you() function